superduperdb/vector_search/postgres.py

```diff
@@ -0,0 +1,146 @@
+import json
+import typing as t
+import numpy
+from pgvector.psycopg import psycopg, register_vector
+
+
+from superduperdb.vector_search.base import BaseVectorSearcher, VectorItem
+
+
+class PostgresVectorSearcher(BaseVectorSearcher):
+    """
+    Implementation of a vector index using the ``pgvector`` library.
+    :param identifier: Unique string identifier of index
+    :param dimensions: Dimension of the vector embeddings in the Lance dataset
+    :param uri: connection string to postgres
+    :param h: ``torch.Tensor``
+    :param index: list of IDs
+    :param measure: measure to assess similarity
+    """
+
+    def __init__(
+        self,
+        identifier: str,
+        dimensions: int,
+        conninfo: str,
+        h: t.Optional[numpy.ndarray] = None,
+        index: t.Optional[t.List[str]] = None,
+        measure: t.Optional[str] = None,
+    ):
+        self.connection = psycopg.connect(conninfo=conninfo)
+        self.dimensions = dimensions
+        self.identifier = identifier
+        if measure == "l2" or not measure:
+            self.measure_query = "embedding <-> '%s'"
+        elif measure == "dot":
+            self.measure_query = "(embedding <#> '%s') * -1"
+        elif measure == "cosine":
+            self.measure_query = "1 - (embedding <=> '%s')"
+        else:
+            raise NotImplementedError("Unrecognized measure format")
+        with self.connection.cursor() as cursor:
+            cursor.execute('CREATE EXTENSION IF NOT EXISTS vector')
+            cursor.execute(
+                'CREATE TABLE IF NOT EXISTS %s (id varchar, embedding vector(%d))'
+                % (self.identifier, self.dimensions)
+            )
+        register_vector(self.connection)
+        if h:
+            self._create_or_append_to_dataset(h, index)
```

Comment:
Every time the current program starts and instantiates it, h may be a loaded vector, so it needs to be checked whether it is built repeatedly. Can you see if this is the case?

Reply:
@jieguangzhou I don't understand your point. Isn't this handled with the `IF EXISTS` statement?

Reply:
@makkarss929 I think @jieguangzhou's concern may be that, we don't want to keep appending to the table multiple times the same data.

Reply:
One possibility is to look at the `ids` and insert only the new `ids`.

Reply:
In my recent commit, declaring `ids` as the primary key to avoid duplication.

Reply:
@makkarss929 Will it throw an error or ignore if the id is duplicated during insertion?